### PR TITLE
Add NPS survey notice and dialog

### PR DESCRIPTION
**Pull request by mattsherman**
This PR implements the flow ([as previously mocked](http://wydzwf.axshare.com)) for prompting the user to take an NPS survey.

nps survey notice

Styling will be added in separate PRs, as will hooking up up the survey block to store responses.

For now, this is not enabled in any environments. To enable:

```
ENABLE_FEATURES=nps-survey/notice make run
```

To re-trigger the NPS survey dialog without reloading the entire page, in the browser console, type:

```
dispatch( { type: "NPS_SURVEY_DIALOG_IS_SHOWING", isShowing: true } );
```

To run unit tests:

```
npm run test-client -- --grep "state ui nps-survey-notice"
```

### CODE REVIEW

File: client/layout/nps-survey-notice/index.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		// wait a little bit before showing the notice, so that
+		// (1) the user gets a chance to look briefly at the uncluttered screen, and
+		// (2) the user notices the notice more, since it will cause a change to the
+		//     screen they are already looking at
+		setTimeout( () => {
+			this.props.showNpsSurveyNotice();
+		}, 2000 );
+	}
+
+	handleDialogClose = () => {
```
Comment (jblz) [inline client/layout/nps-survey-notice/index.jsx]: I'd separate these instance property functions from the rest of the function declarations. To the top of the scope, for example.

File: client/state/ui/nps-survey-notice/test/reducer.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		} );
+
+		it( 'should track if the dialog is showing', () => {
+			const state = isNpsSurveyDialogShowing( undefined, {
+				type: NPS_SURVEY_DIALOG_IS_SHOWING,
+				isShowing: true
+			} );
+
+			expect( state ).to.be.true;
+		} );
```
Comment (jblz) [inline client/state/ui/nps-survey-notice/test/reducer.js]: Is it worth adding a test for dismissing / setting `isShowing: false` here as well?

File: client/layout/nps-survey-notice/index.jsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+class NpsSurveyNotice extends Component {
+	componentDidMount() {
+		// wait a little bit before showing the notice, so that
+		// (1) the user gets a chance to look briefly at the uncluttered screen, and
+		// (2) the user notices the notice more, since it will cause a change to the
+		//     screen they are already looking at
+		setTimeout( () => {
+			this.props.showNpsSurveyNotice();
+		}, 2000 );
```
Comment (jblz) [inline client/layout/nps-survey-notice/index.jsx]: I'd probably push this out about another second -- it pops up a bit quickly for me (even assuming a server round-trip prior)

File: client/layout/nps-survey-notice/index.jsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import { showNpsSurveyNotice, setNpsSurveyDialogShowing } from 'state/ui/nps-survey-notice/actions';
+import { isNpsSurveyDialogShowing } from 'state/ui/nps-survey-notice/selectors';
+
+class NpsSurveyNotice extends Component {
+	componentDidMount() {
+		// wait a little bit before showing the notice, so that
+		// (1) the user gets a chance to look briefly at the uncluttered screen, and
+		// (2) the user notices the notice more, since it will cause a change to the
+		//     screen they are already looking at
+		setTimeout( () => {
```
Comment (jblz) [inline client/layout/nps-survey-notice/index.jsx]: This can be written more concisely as `setTimeout( this.props.showNpsSurveyNotice, 2000 );`

Comment (mattsherman) [inline client/layout/nps-survey-notice/index.jsx]: But of course! :blush:

**Comment by mattsherman** [comment]
> There is a usability question / concern I wanted to discuss, though:
I'd expect I could click anywhere on the notice to initiate the survey dialog (vs. just the Sure! button) -- except the x, of course.

> What do you think? This could be done in a separate PR, I just wanted to throw it out there.

For sure -- that's actually how I had it in the mockup. Unfortunately, the existing global notice component doesn't support that. So, I went with the existing UX convention in Calypso.

This is definitely something we could work on in a future PR though.

**Comment by jblz** [comment]
> I went with the existing UX convention in Calypso.
> This is definitely something we could work on in a future PR though.

Sounds good. :shipit:!

**Comment by mattsherman** [comment]
@jblz Thanks for the review -- I have addressed all of your feedback. It's ready for a final quick review when you get a chance.

**Pull request merged**